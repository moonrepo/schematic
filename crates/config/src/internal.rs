use crate::config::PartialConfig;
use crate::errors::ConfigError;
use crate::merge::merge_partial;
use schematic_types::SchemaType;
use std::{env, str::FromStr};

pub fn default_from_env_var<T: FromStr>(key: &str) -> Result<Option<T>, ConfigError> {
    parse_from_env_var(key, |var| parse_value(var).map(|v| Some(v)))
}

pub fn parse_from_env_var<T>(
    key: &str,
    parser: impl Fn(String) -> Result<Option<T>, ConfigError>,
) -> Result<Option<T>, ConfigError> {
    if let Ok(var) = env::var(key) {
        let value =
            parser(var).map_err(|e| ConfigError::InvalidEnvVar(key.to_owned(), e.to_string()))?;

        return Ok(value);
    }

    Ok(None)
}

pub fn parse_value<T: FromStr, V: AsRef<str>>(value: V) -> Result<T, ConfigError> {
    let value = value.as_ref();

    value.parse::<T>().map_err(|_| {
        ConfigError::Message(format!(
            "Failed to parse \"{value}\" into the correct type."
        ))
    })
}

#[allow(clippy::unnecessary_unwrap)]
pub fn merge_setting<T, C>(
    prev: Option<T>,
    next: Option<T>,
    context: &C,
    merger: impl Fn(T, T, &C) -> Result<Option<T>, ConfigError>,
) -> Result<Option<T>, ConfigError> {
    if prev.is_some() && next.is_some() {
        merger(prev.unwrap(), next.unwrap(), context)
    } else if next.is_some() {
        Ok(next)
    } else {
        Ok(prev)
    }
}

#[allow(clippy::unnecessary_unwrap)]
pub fn merge_partial_setting<T: PartialConfig>(
    prev: Option<T>,
    next: Option<T>,
    context: &T::Context,
) -> Result<Option<T>, ConfigError> {
    if prev.is_some() && next.is_some() {
        merge_partial(prev.unwrap(), next.unwrap(), context)
    } else if next.is_some() {
        // For optional nested configs, defaults values are not provided by the
        // `default_values` method since we return `None`. However, if a value
        // is provided during the merging process, then we need to lazily load
        // the default values to merge against.
        merge_partial(T::default_values(context)?, next.unwrap(), context)
    } else {
        Ok(prev)
    }
}

pub fn partialize_schema(schema: &mut SchemaType) {
    use schematic_types::*;

    match schema {
        SchemaType::Array(ArrayType { items_type, .. }) => {
            partialize_schema(items_type);
        }
        SchemaType::Object(ObjectType {
            key_type,
            value_type,
            ..
        }) => {
            partialize_schema(key_type);
            partialize_schema(value_type);
        }
        SchemaType::Struct(inner) => {
            if let Some(name) = &inner.name {
                inner.name = Some(format!("Partial{}", name));
            }

            inner.partial = true;

            for field in inner.fields.iter_mut() {
                field.optional = true;
                field.nullable = true;

                partialize_schema(&mut field.type_of);

                field.type_of = SchemaType::nullable(field.type_of.clone());
            }
        }
        SchemaType::Tuple(TupleType { items_types, .. }) => {
            for item in items_types {
                partialize_schema(item);
            }
        }
        SchemaType::Union(UnionType {
            variants_types,
            variants,
            ..
        }) => {
            for variant in variants_types {
                partialize_schema(variant);
            }

            if let Some(fields) = variants {
                for field in fields.iter_mut() {
                    field.optional = true;
                    field.nullable = true;

                    partialize_schema(&mut field.type_of);
                }
            }
        }
        _ => {}
    };
}
