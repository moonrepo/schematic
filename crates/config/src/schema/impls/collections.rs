// use crate::config::ConfigSchema;
// use crate::schema::types::*;

// impl<T: ConfigSchema> ConfigSchema for Vec<T> {
//     fn generate_schema() -> Schema {
//         let inner = T::generate_schema();

//         if let Schema::Type {
//             ref mut nullable, ..
//         } = schema
//         {
//             *nullable = true;
//         }

//         schema
//     }
// }
