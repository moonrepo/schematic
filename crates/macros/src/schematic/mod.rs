use crate::common::Macro;
use crate::utils::extract_comment;
use proc_macro2::TokenStream;
use quote::{quote, ToTokens};

pub struct SchematicMacro<'l>(pub Macro<'l>);

impl<'l> ToTokens for SchematicMacro<'l> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let cfg = &self.0;
        let name = cfg.name;
        let schema = cfg.type_of.generate_schema(
            name,
            extract_comment(&cfg.attrs),
            cfg.get_casing_format(),
            cfg.get_tagged_format(),
        );

        tokens.extend(quote! {
            #[automatically_derived]
            impl schematic::Schematic for #name {
                fn generate_schema() -> schematic::SchemaType {
                    use schematic::schema::*;
                    #schema
                }
            }
        });
    }
}
