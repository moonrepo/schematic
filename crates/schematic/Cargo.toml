[package]
name = "schematic"
version = "0.15.1"
edition = "2021"
license = "MIT"
description = "A layered serde configuration and schema library."
homepage = "https://moonrepo.github.io/schematic"
repository = "https://github.com/moonrepo/schematic"
readme = "../../README.md"

[package.metadata.release]
pre-release-replacements = [
	{ file = "../../CHANGELOG.md", search = "Unreleased", replace = "{{version}}" },
]

[package.metadata.docs.rs]
all-features = true

[dependencies]
schematic_macros = { version = "0.15.0", path = "../macros" }
schematic_types = { version = "0.7.0", path = "../types" }
miette = { workspace = true }
thiserror = "1.0.59"
tracing = "0.1.40"

# config
garde = { version = "0.18.0", default-features = false, optional = true, features = [
	"regex",
] }
serde = { workspace = true }
serde_path_to_error = { version = "0.1.16", optional = true }
starbase_styles = { version = "0.4.0", optional = true }

# schema
indexmap = { workspace = true, optional = true, features = ["serde"] }

# json
serde_json = { workspace = true, optional = true, features = [
	"preserve_order",
] }

# json schema
markdown = { version = "1.0.0-alpha.17", optional = true }
schemars = { version = "0.8.17", optional = true, default-features = false }

# toml
toml = { workspace = true, optional = true }

# yaml
serde_yaml = { workspace = true, optional = true }

# url
reqwest = { workspace = true, optional = true, features = ["blocking"] }

[features]
default = ["config", "url"]
config = [
	"dep:garde",
	"dep:serde_path_to_error",
	"dep:starbase_styles",
	"schematic_macros/config",
]
json = ["dep:serde_json"]
schema = ["dep:indexmap", "schematic_macros/schema"]
toml = ["dep:toml"]
url = ["dep:reqwest"]
yaml = ["dep:serde_yaml"]

renderer_json_schema = ["dep:markdown", "dep:schemars", "json", "schema"]
renderer_template = []
renderer_typescript = ["schema"]

type_chrono = ["schematic_types/chrono"]
type_indexmap = ["schematic_types/indexmap"]
type_regex = ["schematic_types/regex"]
type_relative_path = ["schematic_types/relative_path"]
type_rust_decimal = ["schematic_types/rust_decimal"]
type_semver = ["schematic_types/semver"]
type_serde_json = ["schematic_types/serde_json"]
type_serde_toml = ["schematic_types/serde_toml"]
type_serde_yaml = ["schematic_types/serde_yaml"]
type_url = ["schematic_types/url"]

valid_email = ["garde/email"]
valid_url = ["garde/url"]

[dev-dependencies]
schematic = { path = ".", features = [
	"config",
	"json",
	"schema",
	"toml",
	"renderer_json_schema",
	"renderer_template",
	"renderer_typescript",
	"type_chrono",
	"type_indexmap",
	"type_regex",
	"type_relative_path",
	"type_rust_decimal",
	"type_semver",
	"type_serde_json",
	"type_serde_toml",
	"type_serde_yaml",
	"type_url",
	"url",
	"valid_email",
	"valid_url",
	"yaml",
] }
reqwest = { workspace = true, features = [
	"blocking",
	"rustls-tls-native-roots",
] }
serial_test = "3.1.1"
starbase_sandbox = "0.6.0"

# Types
chrono = { workspace = true }
indexmap = { workspace = true }
regex = { workspace = true }
relative-path = { workspace = true, features = ["serde"] }
rust_decimal = { workspace = true }
semver = { workspace = true, features = ["serde"] }
url = { workspace = true, features = ["serde"] }
