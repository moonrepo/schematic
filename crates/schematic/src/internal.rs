use crate::config::{ConfigError, PartialConfig};
use crate::merge::merge_partial;
use schematic_types::Schema;
use std::{env, str::FromStr};

pub fn default_from_env_var<T: FromStr>(key: &str) -> Result<Option<T>, ConfigError> {
    parse_from_env_var(key, |var| parse_value(var).map(|v| Some(v)))
}

pub fn parse_from_env_var<T>(
    key: &str,
    parser: impl Fn(String) -> Result<Option<T>, ConfigError>,
) -> Result<Option<T>, ConfigError> {
    if let Ok(var) = env::var(key) {
        let value =
            parser(var).map_err(|e| ConfigError::InvalidEnvVar(key.to_owned(), e.to_string()))?;

        return Ok(value);
    }

    Ok(None)
}

pub fn parse_value<T: FromStr, V: AsRef<str>>(value: V) -> Result<T, ConfigError> {
    let value = value.as_ref();

    value.parse::<T>().map_err(|_| {
        ConfigError::Message(format!(
            "Failed to parse \"{value}\" into the correct type."
        ))
    })
}

#[allow(clippy::unnecessary_unwrap)]
pub fn merge_setting<T, C>(
    prev: Option<T>,
    next: Option<T>,
    context: &C,
    merger: impl Fn(T, T, &C) -> Result<Option<T>, ConfigError>,
) -> Result<Option<T>, ConfigError> {
    if prev.is_some() && next.is_some() {
        merger(prev.unwrap(), next.unwrap(), context)
    } else if next.is_some() {
        Ok(next)
    } else {
        Ok(prev)
    }
}

#[allow(clippy::unnecessary_unwrap)]
pub fn merge_partial_setting<T: PartialConfig>(
    prev: Option<T>,
    next: Option<T>,
    context: &T::Context,
) -> Result<Option<T>, ConfigError> {
    if prev.is_some() && next.is_some() {
        merge_partial(prev.unwrap(), next.unwrap(), context)
    } else if next.is_some() {
        Ok(next)
    } else {
        Ok(prev)
    }
}

pub fn partialize_schema(schema: &mut Schema, force_partial: bool) {
    use schematic_types::*;

    let mut update_name = |update: bool| {
        if update {
            if let Some(name) = &schema.name {
                if !name.starts_with("Partial") {
                    schema.name = Some(format!("Partial{name}"));
                }
            }
        }
    };

    match &mut schema.ty {
        SchemaType::Array(inner) => {
            partialize_schema(&mut inner.items_type, false);
        }
        SchemaType::Object(inner) => {
            partialize_schema(&mut inner.key_type, false);
            partialize_schema(&mut inner.value_type, false);
        }
        SchemaType::Struct(inner) => {
            if inner.partial || force_partial {
                update_name(true);

                for field in inner.fields.values_mut() {
                    field.optional = true;
                    field.nullify();

                    partialize_schema(field, true);
                }
            } else {
                for field in inner.fields.values_mut() {
                    partialize_schema(field, false);
                }
            }
        }
        SchemaType::Tuple(inner) => {
            for item in inner.items_types.iter_mut() {
                partialize_schema(item, false);
            }
        }
        SchemaType::Union(inner) => {
            update_name(inner.partial || force_partial);

            for variant in inner.variants_types.iter_mut() {
                partialize_schema(variant, false);
            }
        }
        _ => {}
    };
}
