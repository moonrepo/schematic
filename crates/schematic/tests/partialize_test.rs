#![allow(dead_code)]

use schematic::internal::partialize_schema;
use schematic::{derive_enum, Config, ConfigEnum, SchemaBuilder, Schematic};
use similar::{ChangeTag, TextDiff};
use starbase_sandbox::assert_snapshot;
use std::collections::HashMap;
use std::fmt::Debug;

fn create_diff<T: Schematic>() -> String {
    let mut schema = SchemaBuilder::build_root::<T>();

    let original = format!("{:#?}", schema);

    partialize_schema(&mut schema, true);

    let partial = format!("{:#?}", schema);

    // println!("ORIGINAL:\n{}\n\n", original);
    // println!("PARTIAL:\n{}\n\n", partial);

    let mut diff = String::new();

    for change in TextDiff::from_lines(&original, &partial).iter_all_changes() {
        let sign = match change.tag() {
            ChangeTag::Delete => "üü•",
            ChangeTag::Insert => "üü©",
            ChangeTag::Equal => "‚¨õÔ∏è",
        };

        diff.push_str(&format!("{}{}", sign, change));
    }

    diff
}

#[derive(Config)]
struct Empty {}

#[derive(Config)]
struct Basic {
    field: String,
}

#[derive(Config)]
struct Primitives {
    string: String,
    number: usize,
    boolean: bool,
    float: f32,
    string_opt: Option<String>,
    number_opt: Option<isize>,
    boolean_opt: Option<bool>,
    float_opt: Option<f64>,
}

#[test]
fn primitives() {
    assert_snapshot!(create_diff::<Primitives>());
}

#[derive(Config)]
struct Compounds {
    list: Vec<String>,
    list_opt: Option<Vec<usize>>,
    map: HashMap<String, bool>,
    map_opt: Option<HashMap<isize, f32>>,
}

#[test]
fn compounds() {
    assert_snapshot!(create_diff::<Compounds>());
}

derive_enum!(
    #[derive(Default, ConfigEnum)]
    enum UnitEnum {
        #[default]
        A,
        B,
        C,
    }
);

derive_enum!(
    #[derive(Default, ConfigEnum)]
    enum UnitFallbackEnum {
        #[default]
        A,
        B,
        C,
        #[variant(fallback)]
        Other(String),
    }
);

#[derive(Config)]
enum TupleEnum {
    #[setting(nested)]
    A(Empty),
    #[setting(nested)]
    B(Empty),
}

#[derive(Config)]
struct Enums {
    unit: UnitEnum,
    unit_opt: Option<UnitEnum>,
    fallback: UnitFallbackEnum,
    fallback_opt: Option<UnitFallbackEnum>,
    #[setting(nested)]
    tuple: TupleEnum,
    #[setting(nested)]
    tuple_opt: Option<TupleEnum>,
}

#[test]
fn enums() {
    assert_snapshot!(create_diff::<Enums>());
}

#[derive(Config)]
struct Nested {
    #[setting(nested)]
    field: Basic,
    #[setting(nested)]
    field_opt: Option<Basic>,
}

#[test]
fn nested() {
    assert_snapshot!(create_diff::<Nested>());
}

#[derive(Config)]
struct NestedList {
    #[setting(nested)]
    field: Vec<Basic>,
    #[setting(nested)]
    field_opt: Option<Vec<Basic>>,
}

#[test]
fn nested_list() {
    assert_snapshot!(create_diff::<NestedList>());
}

#[derive(Config)]
struct NestedMap {
    #[setting(nested)]
    field: HashMap<String, Basic>,
    #[setting(nested)]
    field_opt: Option<HashMap<usize, Basic>>,
}

#[test]
fn nested_map() {
    assert_snapshot!(create_diff::<NestedMap>());
}

#[derive(Config)]
#[config(serde(untagged))]
enum Untagged {
    Foo,
    Bar(bool),
    Baz(usize, String),
    #[setting(nested)]
    Qux(Basic),
}

#[test]
fn enum_untagged() {
    assert_snapshot!(create_diff::<Untagged>());
}

#[derive(Config)]
enum ExternalTagged {
    Foo,
    Bar(bool),
    Baz(usize),
    #[setting(nested)]
    Qux(Basic),
}

#[test]
fn enum_external() {
    assert_snapshot!(create_diff::<ExternalTagged>());
}

#[derive(Config)]
#[config(serde(tag = "type"))]
enum InternalTagged {
    Foo,
    Bar(bool),
    Baz(usize),
    #[setting(nested)]
    Qux(Basic),
}

#[test]
fn enum_internal() {
    assert_snapshot!(create_diff::<InternalTagged>());
}

#[derive(Config)]
#[config(serde(tag = "type", content = "content"))]
enum AdjacentTagged {
    Foo,
    Bar(bool),
    Baz(usize),
    #[setting(nested)]
    Qux(Basic),
}

#[test]
fn enum_adjacent() {
    assert_snapshot!(create_diff::<AdjacentTagged>());
}
