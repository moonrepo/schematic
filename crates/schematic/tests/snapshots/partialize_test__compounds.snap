---
source: crates/schematic/tests/partialize_test.rs
expression: "create_diff::<Compounds>()"
---
⬛️Schema {
⬛️    description: None,
⬛️    name: Some(
🟥        "Compounds",
🟩        "PartialCompounds",
⬛️    ),
⬛️    ty: Struct(
⬛️        StructType {
⬛️            fields: {
⬛️                "list": Schema {
⬛️                    description: None,
⬛️                    name: None,
🟥                    ty: Array(
🟥                        ArrayType {
🟥                            contains: None,
🟥                            items_type: Schema {
🟥                                description: None,
🟥                                name: None,
🟥                                ty: String(
🟥                                    StringType {
🟥                                        default: None,
🟥                                        enum_values: None,
🟥                                        format: None,
🟥                                        max_length: None,
🟥                                        min_length: None,
🟥                                        pattern: None,
🟥                                    },
🟥                                ),
🟥                                deprecated: None,
🟥                                env_var: None,
🟥                                hidden: false,
🟥                                nullable: false,
🟥                                optional: false,
🟥                                read_only: false,
🟥                                write_only: false,
🟥                            },
🟥                            max_contains: None,
🟥                            max_length: None,
🟥                            min_contains: None,
🟥                            min_length: None,
🟥                            unique: None,
🟩                    ty: Union(
🟩                        UnionType {
🟩                            default_index: None,
🟩                            partial: false,
🟩                            operator: AnyOf,
🟩                            variants_types: [
🟩                                Schema {
🟩                                    description: None,
🟩                                    name: None,
🟩                                    ty: Array(
🟩                                        ArrayType {
🟩                                            contains: None,
🟩                                            items_type: Schema {
🟩                                                description: None,
🟩                                                name: None,
🟩                                                ty: String(
🟩                                                    StringType {
🟩                                                        default: None,
🟩                                                        enum_values: None,
🟩                                                        format: None,
🟩                                                        max_length: None,
🟩                                                        min_length: None,
🟩                                                        pattern: None,
🟩                                                    },
🟩                                                ),
🟩                                                deprecated: None,
🟩                                                env_var: None,
🟩                                                hidden: false,
🟩                                                nullable: false,
🟩                                                optional: false,
🟩                                                read_only: false,
🟩                                                write_only: false,
🟩                                            },
🟩                                            max_contains: None,
🟩                                            max_length: None,
🟩                                            min_contains: None,
🟩                                            min_length: None,
🟩                                            unique: None,
🟩                                        },
🟩                                    ),
🟩                                    deprecated: None,
🟩                                    env_var: None,
🟩                                    hidden: false,
🟩                                    nullable: false,
🟩                                    optional: false,
🟩                                    read_only: false,
🟩                                    write_only: false,
🟩                                },
🟩                                Schema {
🟩                                    description: None,
🟩                                    name: None,
🟩                                    ty: Null,
🟩                                    deprecated: None,
🟩                                    env_var: None,
🟩                                    hidden: false,
🟩                                    nullable: false,
🟩                                    optional: false,
🟩                                    read_only: false,
🟩                                    write_only: false,
🟩                                },
🟩                            ],
⬛️                        },
⬛️                    ),
⬛️                    deprecated: None,
⬛️                    env_var: None,
⬛️                    hidden: false,
🟥                    nullable: false,
🟥                    optional: false,
🟩                    nullable: true,
🟩                    optional: true,
⬛️                    read_only: false,
⬛️                    write_only: false,
⬛️                },
⬛️                "listOpt": Schema {
⬛️                    description: None,
⬛️                    name: None,
⬛️                    ty: Union(
⬛️                        UnionType {
⬛️                            default_index: None,
⬛️                            partial: false,
⬛️                            operator: AnyOf,
⬛️                            variants_types: [
⬛️                                Schema {
⬛️                                    description: None,
⬛️                                    name: None,
⬛️                                    ty: Array(
⬛️                                        ArrayType {
⬛️                                            contains: None,
⬛️                                            items_type: Schema {
⬛️                                                description: None,
⬛️                                                name: None,
⬛️                                                ty: Integer(
⬛️                                                    IntegerType {
⬛️                                                        default: None,
⬛️                                                        enum_values: None,
⬛️                                                        format: None,
⬛️                                                        kind: Usize,
⬛️                                                        max: None,
⬛️                                                        max_exclusive: None,
⬛️                                                        min: None,
⬛️                                                        min_exclusive: None,
⬛️                                                        multiple_of: None,
⬛️                                                    },
⬛️                                                ),
⬛️                                                deprecated: None,
⬛️                                                env_var: None,
⬛️                                                hidden: false,
⬛️                                                nullable: false,
⬛️                                                optional: false,
⬛️                                                read_only: false,
⬛️                                                write_only: false,
⬛️                                            },
⬛️                                            max_contains: None,
⬛️                                            max_length: None,
⬛️                                            min_contains: None,
⬛️                                            min_length: None,
⬛️                                            unique: None,
⬛️                                        },
⬛️                                    ),
⬛️                                    deprecated: None,
⬛️                                    env_var: None,
⬛️                                    hidden: false,
⬛️                                    nullable: false,
⬛️                                    optional: false,
⬛️                                    read_only: false,
⬛️                                    write_only: false,
⬛️                                },
⬛️                                Schema {
⬛️                                    description: None,
⬛️                                    name: None,
⬛️                                    ty: Null,
⬛️                                    deprecated: None,
⬛️                                    env_var: None,
⬛️                                    hidden: false,
⬛️                                    nullable: false,
⬛️                                    optional: false,
⬛️                                    read_only: false,
⬛️                                    write_only: false,
⬛️                                },
⬛️                            ],
⬛️                        },
⬛️                    ),
⬛️                    deprecated: None,
⬛️                    env_var: None,
⬛️                    hidden: false,
⬛️                    nullable: true,
🟥                    optional: false,
🟩                    optional: true,
⬛️                    read_only: false,
⬛️                    write_only: false,
⬛️                },
⬛️                "map": Schema {
⬛️                    description: None,
⬛️                    name: None,
🟥                    ty: Object(
🟥                        ObjectType {
🟥                            key_type: Schema {
🟥                                description: None,
🟥                                name: None,
🟥                                ty: String(
🟥                                    StringType {
🟥                                        default: None,
🟥                                        enum_values: None,
🟥                                        format: None,
🟥                                        max_length: None,
🟥                                        min_length: None,
🟥                                        pattern: None,
🟥                                    },
🟥                                ),
🟥                                deprecated: None,
🟥                                env_var: None,
🟥                                hidden: false,
🟥                                nullable: false,
🟥                                optional: false,
🟥                                read_only: false,
🟥                                write_only: false,
🟥                            },
🟥                            max_length: None,
🟥                            min_length: None,
🟥                            required: None,
🟥                            value_type: Schema {
🟥                                description: None,
🟥                                name: None,
🟥                                ty: Boolean(
🟥                                    BooleanType {
🟥                                        default: None,
🟥                                    },
🟥                                ),
🟥                                deprecated: None,
🟥                                env_var: None,
🟥                                hidden: false,
🟥                                nullable: false,
🟥                                optional: false,
🟥                                read_only: false,
🟥                                write_only: false,
🟥                            },
🟩                    ty: Union(
🟩                        UnionType {
🟩                            default_index: None,
🟩                            partial: false,
🟩                            operator: AnyOf,
🟩                            variants_types: [
🟩                                Schema {
🟩                                    description: None,
🟩                                    name: None,
🟩                                    ty: Object(
🟩                                        ObjectType {
🟩                                            key_type: Schema {
🟩                                                description: None,
🟩                                                name: None,
🟩                                                ty: String(
🟩                                                    StringType {
🟩                                                        default: None,
🟩                                                        enum_values: None,
🟩                                                        format: None,
🟩                                                        max_length: None,
🟩                                                        min_length: None,
🟩                                                        pattern: None,
🟩                                                    },
🟩                                                ),
🟩                                                deprecated: None,
🟩                                                env_var: None,
🟩                                                hidden: false,
🟩                                                nullable: false,
🟩                                                optional: false,
🟩                                                read_only: false,
🟩                                                write_only: false,
🟩                                            },
🟩                                            max_length: None,
🟩                                            min_length: None,
🟩                                            required: None,
🟩                                            value_type: Schema {
🟩                                                description: None,
🟩                                                name: None,
🟩                                                ty: Boolean(
🟩                                                    BooleanType {
🟩                                                        default: None,
🟩                                                    },
🟩                                                ),
🟩                                                deprecated: None,
🟩                                                env_var: None,
🟩                                                hidden: false,
🟩                                                nullable: false,
🟩                                                optional: false,
🟩                                                read_only: false,
🟩                                                write_only: false,
🟩                                            },
🟩                                        },
🟩                                    ),
🟩                                    deprecated: None,
🟩                                    env_var: None,
🟩                                    hidden: false,
🟩                                    nullable: false,
🟩                                    optional: false,
🟩                                    read_only: false,
🟩                                    write_only: false,
🟩                                },
🟩                                Schema {
🟩                                    description: None,
🟩                                    name: None,
🟩                                    ty: Null,
🟩                                    deprecated: None,
🟩                                    env_var: None,
🟩                                    hidden: false,
🟩                                    nullable: false,
🟩                                    optional: false,
🟩                                    read_only: false,
🟩                                    write_only: false,
🟩                                },
🟩                            ],
⬛️                        },
⬛️                    ),
⬛️                    deprecated: None,
⬛️                    env_var: None,
⬛️                    hidden: false,
🟥                    nullable: false,
🟥                    optional: false,
🟩                    nullable: true,
🟩                    optional: true,
⬛️                    read_only: false,
⬛️                    write_only: false,
⬛️                },
⬛️                "mapOpt": Schema {
⬛️                    description: None,
⬛️                    name: None,
⬛️                    ty: Union(
⬛️                        UnionType {
⬛️                            default_index: None,
⬛️                            partial: false,
⬛️                            operator: AnyOf,
⬛️                            variants_types: [
⬛️                                Schema {
⬛️                                    description: None,
⬛️                                    name: None,
⬛️                                    ty: Object(
⬛️                                        ObjectType {
⬛️                                            key_type: Schema {
⬛️                                                description: None,
⬛️                                                name: None,
⬛️                                                ty: Integer(
⬛️                                                    IntegerType {
⬛️                                                        default: None,
⬛️                                                        enum_values: None,
⬛️                                                        format: None,
⬛️                                                        kind: Isize,
⬛️                                                        max: None,
⬛️                                                        max_exclusive: None,
⬛️                                                        min: None,
⬛️                                                        min_exclusive: None,
⬛️                                                        multiple_of: None,
⬛️                                                    },
⬛️                                                ),
⬛️                                                deprecated: None,
⬛️                                                env_var: None,
⬛️                                                hidden: false,
⬛️                                                nullable: false,
⬛️                                                optional: false,
⬛️                                                read_only: false,
⬛️                                                write_only: false,
⬛️                                            },
⬛️                                            max_length: None,
⬛️                                            min_length: None,
⬛️                                            required: None,
⬛️                                            value_type: Schema {
⬛️                                                description: None,
⬛️                                                name: None,
⬛️                                                ty: Float(
⬛️                                                    FloatType {
⬛️                                                        default: None,
⬛️                                                        enum_values: None,
⬛️                                                        format: None,
⬛️                                                        kind: F32,
⬛️                                                        max: None,
⬛️                                                        max_exclusive: None,
⬛️                                                        min: None,
⬛️                                                        min_exclusive: None,
⬛️                                                        multiple_of: None,
⬛️                                                        name: None,
⬛️                                                    },
⬛️                                                ),
⬛️                                                deprecated: None,
⬛️                                                env_var: None,
⬛️                                                hidden: false,
⬛️                                                nullable: false,
⬛️                                                optional: false,
⬛️                                                read_only: false,
⬛️                                                write_only: false,
⬛️                                            },
⬛️                                        },
⬛️                                    ),
⬛️                                    deprecated: None,
⬛️                                    env_var: None,
⬛️                                    hidden: false,
⬛️                                    nullable: false,
⬛️                                    optional: false,
⬛️                                    read_only: false,
⬛️                                    write_only: false,
⬛️                                },
⬛️                                Schema {
⬛️                                    description: None,
⬛️                                    name: None,
⬛️                                    ty: Null,
⬛️                                    deprecated: None,
⬛️                                    env_var: None,
⬛️                                    hidden: false,
⬛️                                    nullable: false,
⬛️                                    optional: false,
⬛️                                    read_only: false,
⬛️                                    write_only: false,
⬛️                                },
⬛️                            ],
⬛️                        },
⬛️                    ),
⬛️                    deprecated: None,
⬛️                    env_var: None,
⬛️                    hidden: false,
⬛️                    nullable: true,
🟥                    optional: false,
🟩                    optional: true,
⬛️                    read_only: false,
⬛️                    write_only: false,
⬛️                },
⬛️            },
⬛️            partial: false,
⬛️            required: None,
⬛️        },
⬛️    ),
⬛️    deprecated: None,
⬛️    env_var: None,
⬛️    hidden: false,
⬛️    nullable: false,
⬛️    optional: false,
⬛️    read_only: false,
⬛️    write_only: false,
⬛️}
